### ISSUE TITLE
GitFlow Support

### ISSUE THREAD

**Issue by red2678** [typicode/husky#842]
Will [GitFlow ](https://github.com/nvie/gitflow) actions (ex. `git flow feature publish`) be supported in v5.x?

**Comment by typicode** [comment]
Not using git-flow personally, but I guess nothing in v5 prevents from using it.
Did you have something particular in mind?

**Comment by red2678** [comment]
Awesome. I incorrectly linked you to the original git-flow repo; this is the most recent iteration [gitflow-avh](https://github.com/petervanderdoes/gitflow-avh).  

Quick note, the gitflow-avh version was [merged](https://github.com/petervanderdoes/gitflow-avh/wiki/Installing-on-Windows#git-for-windows) into [Git for Windows](https://gitforwindows.org/) as of version 2.6.4. I am not sure if it matters to Husky that I am using Git For Windows, but I wanted to point it out. The gitflow-avh version is also installable on all other platforms and comes packaged with Ubuntu as of version [18.04](https://github.com/petervanderdoes/gitflow-avh/wiki/Installing-on-Linux,-Unix,-etc.#ubuntu).

I am looking for support of the [default hooks](https://github.com/petervanderdoes/gitflow-avh/wiki/Reference:-Hooks-and-Filters#hook-scripts) emitted from gitflow-avh's actions.  A few examples can be found [here](https://github.com/petervanderdoes/gitflow-avh/tree/develop/hooks)

There is a `pre` and a `post` for just about every gitflow-avh action. 

Example Command: `git flow feature publish` 

This will emit two "hookable" events `pre-flow-feature-publish` and `post-flow-feature-publish` 

All the core gitflow-avh actions would be supported then. So:

- `{post,pre)-flow-feature-{finish,publish,pull,start,track,delete}`
- `{post,pre}-flow-hotfix-{finish,publish,start,delete}`
- `{post,pre}-flow-bugfix-{finish,publish,start,delete}`
- `{post,pre}-flow-release-{finish,publish,start,track,delete}`

This similar issue was opened, and while GitFlow is an older workflow, I think there is a desire for this feature. https://github.com/typicode/husky/issues/123

**Comment by typicode** [comment]
I'm not on Windows right now. But you could give a try with the following commands:

```sh
git init
npm init -y
npm install husky@next -D
npx husky install
# Add one the hook you'd like to test
npx husky add .husky/<pre-flow-...> "touch ok"
# Run the corresponding git flow command
# Check that "ok" file was created
```

**Comment by red2678** [comment]
Done, but it did not work. I did try the above actually before posting here (as well as again rn). I assumed since the events I wanted to hook into were part of  GIT that Husky supported them out of the box. I had to add one step to the test.

Here are the steps I tried:

```
git init
git flow init (accept all defaults)
npm init -y
npm install husky@next -D
npx husky install
npx husky add .husky/pre-flow-feature check "touch ok"
# Run the corresponding git flow feature start command
# Check that "ok" file was created (it was not)
```

Just as a sanity check I took the file (pre-flow-feature-start) created by huksy and manually put it in `~test_project_root\.git\hooks` (removing the call to `husky.sh` and when I ran the flow command the file `ok` was created.

I made this screencast for you. Hope it helps!

https://user-images.githubusercontent.com/1896036/105344038-69109c00-5bb0-11eb-9021-4551eda30534.mp4

**Comment by typicode** [comment]
Thanks for the screencast :+1: 
Do you know if it was supported in husky 4?

**Comment by red2678** [comment]
You're welcome! I always like recording vs typing out tons of text.

As for it working on v4, no I don't know (it probably did though). I did not have much time to dedicate to this issue when I posted it, sorry to just push it all on you like that. Anyway, I figured out the issue :) Now we just need to fix it, potentially. 

## Reproduce The Issue 
`git init` on an empty directory

This will initialize an empty repo and the config file will look like the below (unless you have global or system configs set).

```
[core]
	repositoryformatversion = 0
	filemode = false
	bare = false
	logallrefupdates = true
	symlinks = false
	ignorecase = true
```
Notice the missing core `hooksPath`

Next, we initialize git flow.
```
git flow init
```
This adds needed config information for git flow to work and the config file **now** looks more or less like: 
```
[core]
	repositoryformatversion = 0
	filemode = false
	bare = false
	logallrefupdates = true
	symlinks = false
	ignorecase = true
[gitflow "branch"]
	master = master
	develop = develop
[gitflow "prefix"]
	feature = feature/
	bugfix = bugfix/
	release = release/
	hotfix = hotfix/
	support = support/
	versiontag = 
[gitflow "path"]
	hooks = C:/Users/anthony/Workspaces/testing/.git/hooks

```
Notice the added core `gitflow "path"`

Now we install husky.
```
npm install husky@next -D
npx husky install
```
This adds needed config information for husky to work and the config file **now** looks more or less like: 

```
[core]
	repositoryformatversion = 0
	filemode = false
	bare = false
	logallrefupdates = true
	symlinks = false
	ignorecase = true
	hooksPath = .husky ** <----------- ADDED  ----------- **
[gitflow "branch"]
	master = master
	develop = develop
[gitflow "prefix"]
	feature = feature/
	bugfix = bugfix/
	release = release/
	hotfix = hotfix/
	support = support/
	versiontag = 
[gitflow "path"]
	hooks = C:/Users/anthony/Workspaces/testing/.git/hooks ** <----------- STILL THE SAME ----------- **

```

## The Issue
When husky is installed into an existing project that is using gitflow husky **does not** update the `gitflow "path"` to point to the correct hooks directory (`.husky`  or whatever you defined it to be).

**Note:** Husky _will work_ if it is installed **BEFORE** you setup gitflow.  Why? Because when you install Husky the `core hooksPath`  is set so when you go to setup gitflow it will correctly suggest you use the hooks dir from the config file (`.husky`). 

So the below will work:
```
git init
npm init -y
npm install husky@next -D
npx husky install
git flow init (accept all defaults) <---------- moved down from step 2 ----------
npx husky add .husky/pre-flow-feature-start "touch ok"
# Run the corresponding git flow feature start command
# Check that "ok" file was created (it was!!)
```

## TL;DR - The Fix
Well, I guess the "real" fix is to make it so that when husky is installed it checks for a  `gitflow "path"` in the git config file and updates it to use whatever directory husky is using. How to do that, atm I am not sure. I need to get back to work lol. I will try to take a peek later when time permits. 

You can run `git config "gitflow.path.hooks" .husky` to manually change the git config file to use the correct directory.

Cheers!

**Comment by typicode** [comment]
Awesome! Thanks for the detailed analysis and fix! :slightly_smiling_face:

**Comment by red2678** [comment]
Thanks! I did fork v5 and take a look (congratulations on the release btw). Its simplicity is very impressive!

To the issue. I am not sure that modifying Husky's install process to look for GitFlow is the right approach. 

The command `git config "gitflow.path.hooks" .husky` could be run as you do here during install:

https://github.com/typicode/husky/blob/79830d3d03902b0b8254a04639122a74bf953f9b/src/commands/install.ts#L32

However, I would argue that MOST people are not using GitFlow. So perhaps we handle this with some documentation around using Husky with GitFlow. Maybe it would be better to just tell people to run the command and set their hooks directory?

Before I make PR I wanted to get your input.

**Comment by typicode** [comment]
Thanks :)

I agree, a mention in the doc would be better.

**Comment by red2678** [comment]
Well, PR made #847 but I am getting a build error.

**Comment by typicode** [comment]
It's merged :) Regarding the build error, it's a bug. I'll check the config.

**Comment by red2678** [comment]
Awesome @typicode thanks!
